core/domain/email_manager_test.py

```diff
@@ -426,3 +428,422 @@ def test_record_of_sent_email_is_written_to_datastore(self):
+        # pylint: disable=protected-access
+        email_hash1 = email_manager._generate_hash('recipient_id',
+                                                   'email_subject',
+                                                   'email_html_body1')
+
+        email_hash2 = email_manager._generate_hash('recipient_id',
+                                                   'email_subject',
+                                                   'email_html_body2')
+        self.assertNotEqual(email_hash1, email_hash2)
+
+        email_hash2 = email_manager._generate_hash('recipient_id2',
+                                                   'email_subject',
+                                                   'email_html_body')
+        self.assertNotEqual(email_hash1, email_hash2)
+
+        email_hash2 = email_manager._generate_hash('recipient_id',
+                                                   'email_subject2',
+                                                   'email_html_body')
+        self.assertNotEqual(email_hash1, email_hash2)
+        # pylint: enable=protected-access
+
+class DuplicateEmailTests(test_utils.GenericTestBase):
+    """Test that duplicate emails are not sent"""
+
+    def setUp(self):
+        super(DuplicateEmailTests, self).setUp()
+
+        self.signup(self.NEW_USER_EMAIL, self.NEW_USER_USERNAME)
+        self.new_user_id = self.get_user_id_from_email(self.NEW_USER_EMAIL)
+
+        self.signup(self.ADMIN_EMAIL, self.ADMIN_USERNAME)
+        self.admin_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        self.set_admins([self.ADMIN_USERNAME])
+
+        self.new_footer = (
+            'Unsubscribe from emails at your '
+            '<a href="https://www.site.com/prefs">Preferences page</a>.')
+        self.new_email_content = {
+            'subject': 'Welcome!',
+            'html_body': (
+                'Here is some HTML text.<br>'
+                'With a <b>bold</b> bit and an <i>italic</i> bit.<br>')
+        }
+
+    def test_send_email_does_not_resend_if_same_hash_exists(self):
+        can_send_emails_ctx = self.swap(
+            feconf, 'CAN_SEND_EMAILS_TO_USERS', True)
+
+        duplicate_email_ctx = self.swap(
+            feconf, 'DUPLICATE_EMAIL_INTERVAL_MINS', 1000)
+
+        logged_errors = []
+
+        def _log_error_for_tests(error_message):
+            logged_errors.append(error_message)
+
+        log_new_error_counter = test_utils.CallCounter(_log_error_for_tests)
+        log_new_error_ctx = self.swap(
+            email_manager, 'log_new_error', log_new_error_counter)
+
+        with can_send_emails_ctx, duplicate_email_ctx, log_new_error_ctx:
+            all_models = email_models.SentEmailModel.get_all().fetch()
+            self.assertEqual(len(all_models), 0)
+
+            # pylint: disable=protected-access
+            email_hash = email_manager._generate_hash(self.new_user_id,
+                                                      'Email Subject',
```

Comment:
Just FYI, here (and elsewhere), you can do:

```
           email_hash = email_manager._generate_hash(
               self.new_user_id, 'Email Subject', 'Email Body')	
```

instead of squashing everything all the way over to the right.